packages/common/src/common.ts

```diff
@@ -285,57 +285,89 @@ export class Common extends EventEmitter {
    * will be thrown).
    *
    * @param blockNumber
-   * @param td
+   * @param td : total difficulty of the parent block (for block hf) OR of the the chain latest (for chain hf)
    * @returns The name of the HF
    */
   getHardforkByBlockNumber(blockNumber: BigIntLike, td?: BigIntLike): string {
     blockNumber = toType(blockNumber, TypeOutput.BigInt)
     td = toType(td, TypeOutput.BigInt)
 
-    let hardfork = Hardfork.Chainstart
-    let minTdHF
-    let maxTdHF
-    let previousHF
+    /** The hardfork config seen at hardfork before the merge */
+    let preMergeHF: HardforkConfig | undefined = undefined
+    /**
+     * Merge hardfork config if we have passed the merge hardfork while finding hardfork
+     * because next hardfork's block <= blockNumber
+     */
+    let passedMergeHF: HardforkConfig | undefined = undefined
+    /** Candiate hardfork config, updated as we iterate amongst the hardforks */
+    let hardfork: HardforkConfig | undefined = undefined
+
+    // Hardforks specified are strickly in the order of application, so we can make some intelligent decisions
+    // based on that especially regarding the merge hardfork where block might not be specified
     for (const hf of this.hardforks()) {
-      // Skip comparison for not applied HFs
-      if (hf.block === null) {
-        if (td !== undefined && td !== null && hf.ttd !== undefined && hf.ttd !== null) {
-          if (td >= BigInt(hf.ttd)) {
-            return hf.name
-          }
-        }
+      // Check if hf can be applied, if not then break and the last hardfork set is the response
+      if (hf.block === null && (hf.ttd === undefined || hf.ttd === null)) {
+        // Not a hardfork which has been scheduled yet so just skip
         continue
       }
-      if (blockNumber >= BigInt(hf.block)) {
-        hardfork = hf.name as Hardfork
+
+      if (hf.ttd !== undefined && hf.ttd !== null && preMergeHF === undefined) {
```

Comment:
Shouldn't we throw here if `preMergeHF !== undefined`?

Reply:
ummm...  can chain start from pre merge hardfork specification (i.e. genesi post merge), followed by hard forks still activated at genesis (`0`), just that the user jumbled them up, In long term there can be a dag dependency that we can validate

Reply:
If you find a preMergeHf at this point, doesn't that mean you found a second HF with a ttd specified (which is invalid)?  And if the user jumbled up the hardfork sequence, that's still invalid.  We shouldn't allow hardforks to be out of sequence.

Reply:
no I need to set `preMergeHf` here if this `hf` is the merge one so that if post merge hardforks don't quality based on block numbers, the choice has to be made from `preMergeHf` and the merge hf (which will be set in `hardfork` if the next hardfork don't quality and we break out of the loop)

the check of double ttd hardfork is made  down (towards loop end) as `if (passedMergeHF !== undefined && hf.ttd !== undefined && hf.ttd !== null) {`

Reply:
> We shouldn't allow hardforks to be out of sequence.
Ideally we should have a dag, which should impose partial order on the hardfork specification, but for now let user handle it. The order isn't really fixed for e.g. `mergeForkIdTransition` could come before (like it came in `kiln` network) or could come after `merge` (like sepolia and then other networks will followup)

so its better to apply an open approcah here, and have some sort of dag validation when we are setting the `hardforks` in the common while instantiation.

Reply:
I guess maybe I'm not tracking something then.  This is the first validation step and the first thing you do is check to see if the hardfork being checked as a TTD.  If it does, you check to see if you've set a "preMergeHf" and if not, you set it.  If the preMergeHf already exists, then you have another hf with TTD, which we agree is an invalid case.  My point is that I think this validation step takes care of the "no two hardforks with TTD allowed" and you don't need a separate check.

Reply:
ohh thanks for explaing , yes it should, will update and see if test file runs

Reply:
yes it works! updated